File: Utilities/Queue.rb
# Should make a node class and use that for each data point instead of the array

class Queue
    attr_accessor(:data)
    def initialize()
      @data = []
    end

    def add(value)
        self.data.push(value)
    end

    def remove()
        return self.data.shift()
    end

    def peek()
        return self.data[0]
    end

    def is_empty()
        self.peek() == nil ? return true : return false
    end
end

File: Utilities/Queue.js
class Node {
  constructor(data, next = null) {
    this.data = data
    this.next = next
  }
}

class Queue {
  constructor(first = null, last = null) {
    this.first = first
    this.last = last
  }

  add(data) {
    const currNode = new Node(data)
    if (this.last !== null) {
      this.last.next = currNode
    }
    this.last = currNode
    if (this.first === null) {
      this.first = this.last
    }
  }

  remove() {
    const currNode = this.first
    const first = currNode.next
    if (first === null) {
      this.last = null
    }
    return currNode.data
  }

  peek() {
    return this.first.data
  }

  isEmpty() {
    if (this.peek() === null) {
      return true
    } else {
      return false
    }
  }
}

exports.Queue = Queue
